### PR TITLE
@ckeditor/ckeditor5-core improve PluginCollection.get() typing

### PR DESCRIPTION
**Pull request by bendemboski**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/ckeditor/ckeditor5/blob/1e5b03d8a91038dfcf4a7afc7a47f11cb7a27041/packages/ckeditor5-core/src/plugincollection.js#L110
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by fedemp** [comment]
I'm quoting @bendemboski from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54705

> `let somePlugin = editor.plugins.get(SomePlugin.pluginName) as SomePlugin`

````ts
// This is what I would do
let somePlugin = editor.plugins.get(SomePlugin,pluginName); // You get a `Plugin`.
if (somePlugin instanceof SomePlugin) {
    somePlugin.someMethod();
} 
````

> `type PluginConstructor<T = Plugin> = { new (editor: Editor): T }`

Wouldn't this forbid you from doing `SomePlugin(editor, options)`?

**Comment by bendemboski** [comment]
> Wouldn't this forbid you from doing SomePlugin(editor, options)?

Plugins don't get options passed to them -- all their options are read in from the editor config, and they are only passed an `Editor` instance. See [Plugin constructor](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_plugin-Plugin.html#function-constructor) and [plugin instantiation code](https://github.com/ckeditor/ckeditor5/blob/8b58f1c180d8a862de9d5d9bc8561a8c0925f617/packages/ckeditor5-core/src/plugincollection.js#L421).

**Comment by fedemp** [comment]
> > Wouldn't this forbid you from doing SomePlugin(editor, options)?
> 
> Plugins don't get options passed to them -- all their options are read in from the editor config, and they are only passed an `Editor` instance. See [Plugin constructor](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_plugin-Plugin.html#function-constructor) and [plugin instantiation code](https://github.com/ckeditor/ckeditor5/blob/8b58f1c180d8a862de9d5d9bc8561a8c0925f617/packages/ckeditor5-core/src/plugincollection.js#L421).

You are right. I mixed them with `Command`, that I'm typing right now.

File: types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
Comment (fedemp) [inline types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts:59]: Relying in a string can lead to errors. As I mentioned in the comments, you should check if your variable is `instanceof` a specific `Plugin` subclass.

**Comment by fedemp** [comment]
@bendemboski Please check this PR that is awaiting review.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54858/files

File: types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
Comment (bendemboski) [reply types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts:59]: I entirely agree that using the form of `get()` that accepts a string is inherently type-unsafe, and will always need to be coupled with a type casting, or an `instanceof` check (which I also agree is safer than just type casting).

Using the form of `get()` that accepts a plugin class (constructor) reliably returns an instance of that class, so no `instanceof` check or casting is needed in that case, which is the main use case this PR is enabling.

Comment (bendemboski) [reply types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts:59]: I do think you were squashing some problems with the types by using `instanceof`. That's a strong enough check that the typescript compiler will let you treat the value as an instance of the plugin inside the if block, but the type is still incorrect -- e.g. try actually casting the return value to an instance of your plugin class. When I tried the compiler complained that `MyPlugin` and `() => Plugin` don't overlap sufficiently.

**Comment by bendemboski** [comment]
@fedemp ah, nice, there's some great stuff in there! It looks like it will still require a type casting for the plugin class form. In other words, you can't just do this:

```ts
editor.plugins.get(MyPlugin).myCMethod();
```

but will instead need to type cast or use an instanceof check.

I'd be happy to wait until that PR is merged and then rebase/update my PR, or if you like, you can just pull my `get()` changes into your PR to enable the above scenario and I'll close this one.

Thoughts?

**Comment by fedemp** [comment]
Please let's merge your PR first. You deserve the recognition for the idea. :slightly_smiling_face: 

Aside from adding support for `ContextPlugin`, I think we reuse the `PluginInterface` that is already declared in CKEditor.

````diff
diff --git c/types/ckeditor__ckeditor5-core/src/plugin.d.ts i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
index 5ddde725bb..71a2e1fa1e 100644
--- c/types/ckeditor__ckeditor5-core/src/plugin.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
@@ -45,25 +45,17 @@ export default abstract class Plugin implements Emitter, Observable {
     set(name: string, value: unknown): void;
     set(option: Record<string, string>): void;
     stopDelegating(event?: string, emitter?: Emitter): void;
-    stopListening(
-        emitter?: Emitter,
-        event?: string,
-        callback?: (info: EventInfo, data: DomEventData) => void,
-    ): void;
+    stopListening(emitter?: Emitter, event?: string, callback?: (info: EventInfo, data: DomEventData) => void): void;
     unbind(...unbindProperties: string[]): void;
 }
 
 // Beware that this defines a class constructor, not the class instance.
 // These readonly properties are static properties.
-export interface PluginInterface {
-    new (editor: Editor): {
-        init?(): Promise<void> | void;
-        afterInit?(): Promise<void> | void;
-        destroy?(): Promise<void> | void;
-    };
+export interface PluginInterface<T = Plugin> {
+    new (editor: Editor): T;
     readonly pluginName?: string | undefined;
     readonly isContextPlugin?: boolean | undefined;
     readonly requires?: Array<typeof Plugin | typeof ContextPlugin | string> | undefined;
 }
 
-export type LoadedPlugins = Array<typeof Plugin|typeof ContextPlugin>;
+export type LoadedPlugins = Array<typeof Plugin | typeof ContextPlugin>;
diff --git c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
index 7ab754ff63..083ff712c3 100644
--- c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
@@ -4,11 +4,7 @@ import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
 import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
 import Context from "./context";
 import Editor from "./editor/editor";
-import Plugin, { LoadedPlugins } from "./plugin";
-
-interface PluginConstructor<T = Plugin> {
-    new (editor: Editor): T;
-}
+import Plugin, { LoadedPlugins, PluginInterface } from "./plugin";
 
 export default class PluginCollection implements Emitter, Iterable<[typeof Plugin, Plugin]> {
     constructor(
@@ -20,10 +16,10 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     [Symbol.iterator](): Iterator<[typeof Plugin, Plugin]>;
     destroy(): Promise<void>;
 
-    get<T extends Plugin>(key: PluginConstructor<T>): T;
+    get<T extends Plugin>(key: PluginInterface<T>): T;
     get(key: string): Plugin;
 
-    has(key: PluginConstructor | string): boolean;
+    has(key: PluginInterface | string): boolean;
 
     init(
         plugins: Array<(() => Plugin) | string>,
@@ -53,5 +49,3 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     delegate(...events: string[]): EmitterMixinDelegateChain;
     stopDelegating(event?: string, emitter?: Emitter): void;
 }
-
-export {};
````

**Comment by fedemp** [comment]
Even better diff...

````diff
diff --git c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
index 2c063a2d37..53359b4759 100644
--- c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
+++ i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
@@ -76,6 +76,9 @@ const myPlugin2 = editor.plugins.get(MyPlugin);
 myPlugin2.myMethod() === null;
 const myPlugin3 = editor.plugins.get('MyPlugin') as MyPlugin;
 myPlugin3.myMethod() === null;
+// $ExpectError
+editor.plugins.get(class Foo);
+editor.plugins.get(class Foo extends Plugin {});
 
 editor.plugins.has(MyPlugin);
 editor.plugins.has('MyPlugin');
diff --git c/types/ckeditor__ckeditor5-core/src/plugin.d.ts i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
index 5ddde725bb..9e6593073b 100644
--- c/types/ckeditor__ckeditor5-core/src/plugin.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
@@ -45,25 +45,14 @@ export default abstract class Plugin implements Emitter, Observable {
     set(name: string, value: unknown): void;
     set(option: Record<string, string>): void;
     stopDelegating(event?: string, emitter?: Emitter): void;
-    stopListening(
-        emitter?: Emitter,
-        event?: string,
-        callback?: (info: EventInfo, data: DomEventData) => void,
-    ): void;
+    stopListening(emitter?: Emitter, event?: string, callback?: (info: EventInfo, data: DomEventData) => void): void;
     unbind(...unbindProperties: string[]): void;
 }
 
 // Beware that this defines a class constructor, not the class instance.
 // These readonly properties are static properties.
-export interface PluginInterface {
-    new (editor: Editor): {
-        init?(): Promise<void> | void;
-        afterInit?(): Promise<void> | void;
-        destroy?(): Promise<void> | void;
-    };
-    readonly pluginName?: string | undefined;
-    readonly isContextPlugin?: boolean | undefined;
-    readonly requires?: Array<typeof Plugin | typeof ContextPlugin | string> | undefined;
+export interface PluginInterface<T extends Plugin = Plugin> {
+    new (editor: Editor): T;
 }
 
-export type LoadedPlugins = Array<typeof Plugin|typeof ContextPlugin>;
+export type LoadedPlugins = Array<typeof Plugin | typeof ContextPlugin>;
diff --git c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
index 7ab754ff63..083ff712c3 100644
--- c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
@@ -4,11 +4,7 @@ import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
 import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
 import Context from "./context";
 import Editor from "./editor/editor";
-import Plugin, { LoadedPlugins } from "./plugin";
-
-interface PluginConstructor<T = Plugin> {
-    new (editor: Editor): T;
-}
+import Plugin, { LoadedPlugins, PluginInterface } from "./plugin";
 
 export default class PluginCollection implements Emitter, Iterable<[typeof Plugin, Plugin]> {
     constructor(
@@ -20,10 +16,10 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     [Symbol.iterator](): Iterator<[typeof Plugin, Plugin]>;
     destroy(): Promise<void>;
 
-    get<T extends Plugin>(key: PluginConstructor<T>): T;
+    get<T extends Plugin>(key: PluginInterface<T>): T;
     get(key: string): Plugin;
 
-    has(key: PluginConstructor | string): boolean;
+    has(key: PluginInterface | string): boolean;
 
     init(
         plugins: Array<(() => Plugin) | string>,
@@ -53,5 +49,3 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     delegate(...events: string[]): EmitterMixinDelegateChain;
     stopDelegating(event?: string, emitter?: Emitter): void;
 }
-
-export {};
````

**Comment by fedemp** [comment]
Added support for `ContextPlugin`.

```diff
diff --git c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
index 2c063a2d37..33e59d7ef0 100644
--- c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
+++ i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
@@ -76,6 +76,9 @@ const myPlugin2 = editor.plugins.get(MyPlugin);
 myPlugin2.myMethod() === null;
 const myPlugin3 = editor.plugins.get('MyPlugin') as MyPlugin;
 myPlugin3.myMethod() === null;
+// $ExpectError
+editor.plugins.get(class Foo);
+editor.plugins.get(class Foo extends Plugin {});
 
 editor.plugins.has(MyPlugin);
 editor.plugins.has('MyPlugin');
@@ -128,16 +131,29 @@ contextWithConfig.initPlugins().then(plugins => plugins.map(plugin => plugin.plu
 /**
  * ContextPlugin
  */
-const CPlugin = new ContextPlugin(context) && new ContextPlugin(editor);
-const afterInitPromise = CPlugin.afterInit?.();
+class MyContextPlugin extends ContextPlugin {
+    builtinPlugins: [MyPlugin];
+    contextMethod(): null {
+        return null;
+    }
+}
+let myContextPlugin = new MyContextPlugin(context) && new MyContextPlugin(editor);
+myContextPlugin.contextMethod() === null;
+const afterInitPromise = myContextPlugin.afterInit?.();
 if (afterInitPromise != null) {
     afterInitPromise.then(() => {});
 }
 
-class MyCPlugin extends ContextPlugin {
-    builtinPlugins: [MyPlugin];
+myContextPlugin = editor.plugins.get(MyContextPlugin)
+myContextPlugin.contextMethod() === null;
+
+const unknownPlugin = editor.plugins.get('myContextPlugin');
+if (unknownPlugin instanceof MyContextPlugin) {
+    unknownPlugin.contextMethod() === null;
 }
 
+editor.plugins.has(MyContextPlugin) === bool;
+
 /**
  * DataApiMixin
  */
diff --git c/types/ckeditor__ckeditor5-core/diff i/types/ckeditor__ckeditor5-core/diff
new file mode 100644
index 0000000000..b63847d422
--- /dev/null
+++ i/types/ckeditor__ckeditor5-core/diff
@@ -0,0 +1,84 @@
+diff --git c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
+index 2c063a2d37..53359b4759 100644
+--- c/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
++++ i/types/ckeditor__ckeditor5-core/ckeditor__ckeditor5-core-tests.ts
+@@ -76,6 +76,9 @@ const myPlugin2 = editor.plugins.get(MyPlugin);
+ myPlugin2.myMethod() === null;
+ const myPlugin3 = editor.plugins.get('MyPlugin') as MyPlugin;
+ myPlugin3.myMethod() === null;
++// $ExpectError
++editor.plugins.get(class Foo);
++editor.plugins.get(class Foo extends Plugin {});
+ 
+ editor.plugins.has(MyPlugin);
+ editor.plugins.has('MyPlugin');
+diff --git c/types/ckeditor__ckeditor5-core/src/plugin.d.ts i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
+index 5ddde725bb..9e6593073b 100644
+--- c/types/ckeditor__ckeditor5-core/src/plugin.d.ts
++++ i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
+@@ -45,25 +45,14 @@ export default abstract class Plugin implements Emitter, Observable {
+     set(name: string, value: unknown): void;
+     set(option: Record<string, string>): void;
+     stopDelegating(event?: string, emitter?: Emitter): void;
+-    stopListening(
+-        emitter?: Emitter,
+-        event?: string,
+-        callback?: (info: EventInfo, data: DomEventData) => void,
+-    ): void;
++    stopListening(emitter?: Emitter, event?: string, callback?: (info: EventInfo, data: DomEventData) => void): void;
+     unbind(...unbindProperties: string[]): void;
+ }
+ 
+ // Beware that this defines a class constructor, not the class instance.
+ // These readonly properties are static properties.
+-export interface PluginInterface {
+-    new (editor: Editor): {
+-        init?(): Promise<void> | void;
+-        afterInit?(): Promise<void> | void;
+-        destroy?(): Promise<void> | void;
+-    };
+-    readonly pluginName?: string | undefined;
+-    readonly isContextPlugin?: boolean | undefined;
+-    readonly requires?: Array<typeof Plugin | typeof ContextPlugin | string> | undefined;
++export interface PluginInterface<T extends Plugin = Plugin> {
++    new (editor: Editor): T;
+ }
+ 
+-export type LoadedPlugins = Array<typeof Plugin|typeof ContextPlugin>;
++export type LoadedPlugins = Array<typeof Plugin | typeof ContextPlugin>;
+diff --git c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
+index 7ab754ff63..083ff712c3 100644
+--- c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
++++ i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
+@@ -4,11 +4,7 @@ import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
+ import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
+ import Context from "./context";
+ import Editor from "./editor/editor";
+-import Plugin, { LoadedPlugins } from "./plugin";
+-
+-interface PluginConstructor<T = Plugin> {
+-    new (editor: Editor): T;
+-}
++import Plugin, { LoadedPlugins, PluginInterface } from "./plugin";
+ 
+ export default class PluginCollection implements Emitter, Iterable<[typeof Plugin, Plugin]> {
+     constructor(
+@@ -20,10 +16,10 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
+     [Symbol.iterator](): Iterator<[typeof Plugin, Plugin]>;
+     destroy(): Promise<void>;
+ 
+-    get<T extends Plugin>(key: PluginConstructor<T>): T;
++    get<T extends Plugin>(key: PluginInterface<T>): T;
+     get(key: string): Plugin;
+ 
+-    has(key: PluginConstructor | string): boolean;
++    has(key: PluginInterface | string): boolean;
+ 
+     init(
+         plugins: Array<(() => Plugin) | string>,
+@@ -53,5 +49,3 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
+     delegate(...events: string[]): EmitterMixinDelegateChain;
+     stopDelegating(event?: string, emitter?: Emitter): void;
+ }
+-
+-export {};
diff --git c/types/ckeditor__ckeditor5-core/src/contextplugin.d.ts i/types/ckeditor__ckeditor5-core/src/contextplugin.d.ts
index f9dec1c656..25f9092588 100644
--- c/types/ckeditor__ckeditor5-core/src/contextplugin.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/contextplugin.d.ts
@@ -3,7 +3,7 @@ import Context from "./context";
 import Editor from "./editor/editor";
 import Plugin from "./plugin";
 
-export default class ContextPlugin implements Observable {
+export default abstract class ContextPlugin implements Observable {
     readonly context: Context | Editor;
 
     static readonly isContextPlugin: boolean;
@@ -21,3 +21,7 @@ export default class ContextPlugin implements Observable {
     unbind(...unbindProperties: string[]): void;
     decorate(methodName: string): void;
 }
+
+export interface ContextPluginInterface<T extends ContextPlugin = ContextPlugin> {
+    new (context: Editor | Context): T;
+}
diff --git c/types/ckeditor__ckeditor5-core/src/plugin.d.ts i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
index 5ddde725bb..9e6593073b 100644
--- c/types/ckeditor__ckeditor5-core/src/plugin.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugin.d.ts
@@ -45,25 +45,14 @@ export default abstract class Plugin implements Emitter, Observable {
     set(name: string, value: unknown): void;
     set(option: Record<string, string>): void;
     stopDelegating(event?: string, emitter?: Emitter): void;
-    stopListening(
-        emitter?: Emitter,
-        event?: string,
-        callback?: (info: EventInfo, data: DomEventData) => void,
-    ): void;
+    stopListening(emitter?: Emitter, event?: string, callback?: (info: EventInfo, data: DomEventData) => void): void;
     unbind(...unbindProperties: string[]): void;
 }
 
 // Beware that this defines a class constructor, not the class instance.
 // These readonly properties are static properties.
-export interface PluginInterface {
-    new (editor: Editor): {
-        init?(): Promise<void> | void;
-        afterInit?(): Promise<void> | void;
-        destroy?(): Promise<void> | void;
-    };
-    readonly pluginName?: string | undefined;
-    readonly isContextPlugin?: boolean | undefined;
-    readonly requires?: Array<typeof Plugin | typeof ContextPlugin | string> | undefined;
+export interface PluginInterface<T extends Plugin = Plugin> {
+    new (editor: Editor): T;
 }
 
-export type LoadedPlugins = Array<typeof Plugin|typeof ContextPlugin>;
+export type LoadedPlugins = Array<typeof Plugin | typeof ContextPlugin>;
diff --git c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
index 7ab754ff63..b15c91c4db 100644
--- c/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
+++ i/types/ckeditor__ckeditor5-core/src/plugincollection.d.ts
@@ -4,11 +4,7 @@ import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
 import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
 import Context from "./context";
 import Editor from "./editor/editor";
-import Plugin, { LoadedPlugins } from "./plugin";
-
-interface PluginConstructor<T = Plugin> {
-    new (editor: Editor): T;
-}
+import Plugin, { LoadedPlugins, PluginInterface } from "./plugin";
 
 export default class PluginCollection implements Emitter, Iterable<[typeof Plugin, Plugin]> {
     constructor(
@@ -20,10 +16,11 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     [Symbol.iterator](): Iterator<[typeof Plugin, Plugin]>;
     destroy(): Promise<void>;
 
-    get<T extends Plugin>(key: PluginConstructor<T>): T;
-    get(key: string): Plugin;
+    get<T extends Plugin>(key: PluginInterface<T>): T;
+    get<T extends ContextPlugin>(key: ContextPluginInterface<T>): T;
+    get(key: string): Plugin | ContextPlugin;
 
-    has(key: PluginConstructor | string): boolean;
+    has(key: PluginInterface | ContextPluginInterface | string): boolean;
 
     init(
         plugins: Array<(() => Plugin) | string>,
@@ -53,5 +50,3 @@ export default class PluginCollection implements Emitter, Iterable<[typeof Plugi
     delegate(...events: string[]): EmitterMixinDelegateChain;
     stopDelegating(event?: string, emitter?: Emitter): void;
 }
-
-export {};
```

**Comment by bendemboski** [comment]
@fedemp I re-focused this on only the .get() type, since I think your PR does a better job of improving the .has() type. I'm happy to bring the .has() stuff back into this PR if you prefer, but it seems like maybe it'll be easier to roll it into your PR?

Aside from that, I added some test improvements although not all of the ones you suggested in your diff (the commit crossed the notification). Let me know if there's more test coverage you'd like to see in this PR vs. in yours.

File: types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts
Comment (fedemp) [inline types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts:1]: Is this really needed?

**Review by fedemp — CHANGES_REQUESTED** [review]
Just this unneeded import. I'll approve then.

File: types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts
Comment (bendemboski) [reply types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts:1]: Yeah, this is needed because the tests reference the `Plugin` symbol [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a5e706fab1bc9a59c8797833becad4546f1b5be1/types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts#L34) and without this import, it implicitly references the browser global [Plugin](https://developer.mozilla.org/en-US/docs/Web/API/Plugin). I'm not quite sure why this wasn't a problem before, but probably something to do with the changes to the `PluginInterface` definition caused [this line](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a5e706fab1bc9a59c8797833becad4546f1b5be1/types/ckeditor__ckeditor5-editor-inline/ckeditor__ckeditor5-editor-inline-tests.ts#L41) to start failing with:

Fullscreen_8_5_21__7_34_AM

because `MyPlugin` was a `window.Plugin`, not a CKEditor `Plugin`.

**Review by fedemp — APPROVED** [review]
Great work. Thanks for teaching me some tricks. You should add yourself as an owner if you like. :slightly_smiling_face:

**Comment by bendemboski** [comment]
Haha, you bet! There's so much to learn in typescript, I still feel like a newbie most of the time 😆

Thanks a lot for creating and maintaining these types -- we use CKEditor extensively and as we've been porting our code base to typescript, it's been fantastic having types for CKEditor!

I'd be happy to help maintain. Would that involve opening a PR with

```
// Definitions by: Federico Panico <https://github.com/fedemp>
//                 Ben Demboski <https://github.com/bendemboski>
```

in all the CKEditor package `index.d.ts`s? Or is there some other procedure? (this is my first time DefinitelyTyped'ing)

**Comment by fedemp** [comment]
> Thanks a lot for creating and maintaining these types -- we use CKEditor extensively and as we've been porting our code base to typescript, it's been fantastic having types for CKEditor!

That's the best thing I have read in a long time. :smile: 

About adding yourself as an owner, you are right. Just add yourself in the `index.d.ts` for this package.

**Comment by bendemboski** [comment]
Ready to merge

**Pull request merged**
